domain/access/state/permission.go

```diff
@@ -195,74 +198,167 @@ func (st *PermissionState) UpsertPermission(ctx context.Context, args access.Upd
+			return errors.Annotatef(accesserrors.UserNotFound, "looking for permissions for %q on %q", subject, target.Key)
 		} else if err != nil {
 			return errors.Annotatef(err, "getting permission for %q on %q", subject, target.Key)
 		}
 
+		if user.External {
+			baseExternalPerms, err = st.baseExternalAccessForTarget(ctx, tx, target)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
 		return nil
 	})
 	if err != nil {
-		return userAccess, errors.Trace(err)
+		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
 	}
 
-	return readUser.toUserAccess(permUser), nil
+	userAccess, err := st.generateUserAccess(user, perm, baseExternalPerms)
+	if err != nil {
+		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
+	}
+
+	return userAccess, err
 }
 
 // ReadUserAccessLevelForTarget returns the subject's (user) access level
 // for the given user on the given target.
 func (st *PermissionState) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
-	userAccessType := corepermission.NoAccess
+	userAccess := corepermission.NoAccess
 	db, err := st.DB()
 	if err != nil {
-		return userAccessType, errors.Trace(err)
+		return userAccess, errors.Trace(err)
+	}
+
+	user := dbPermissionUser{
+		Name: subject,
+	}
+	perm := dbPermission{
+		GrantOn: target.Key,
+	}
+
+	readQuery := `
+SELECT  (u.external) AS (&dbPermissionUser.*),
+        (p.access_type, p.uuid) AS (&dbPermission.*)
+FROM    v_user_auth u
+        LEFT JOIN v_permission p ON u.uuid = p.grant_to AND p.grant_on = $dbPermission.grant_on
+WHERE   u.name = $dbPermissionUser.name
+AND     u.disabled = false
+AND     u.removed = false
+`
+
+	readStmt, err := st.Prepare(readQuery, user, perm)
+	if err != nil {
+		return userAccess, errors.Annotatef(err, "preparing select user access level for target statement")
 	}
 
+	var baseExternalPerms dbPermission
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		userAccessType, err = st.userAccessLevel(ctx, tx, subject, target)
+		err = tx.Query(ctx, readStmt, user, perm).Get(&user, &perm)
+		if errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("%w for %q on %q", accesserrors.AccessNotFound, subject, target.Key)
+		} else if err != nil {
+			return errors.Annotatef(err, "reading user access level for target")
+		}
+
+		if user.External {
+			baseExternalPerms, err = st.baseExternalAccessForTarget(ctx, tx, target)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
 		return err
 	})
 	if err != nil {
-		return userAccessType, errors.Trace(err)
+		return userAccess, errors.Trace(domain.CoerceError(err))
+	}
+
+	userAccess = corepermission.Access(perm.AccessType)
+	if user.External && baseExternalAccessGreater(baseExternalPerms, userAccess) {
+		return corepermission.Access(baseExternalPerms.AccessType), nil
+	}
+	if perm.UUID != "" {
+		return userAccess, nil
+	}
```

Comment:
There has to be a better way to signal a valid perm?

Reply:
Changed to `if p.AccessType != string(corepermission.NoAccess) {`. 

I need to do this check because I'm now left joining rather than inner joining the permissions and users table, since even if a user has no permissions, they may be external and have everyone@external permissions. So if they do really have no permissions, we still get a row, its just got no permission data in it.

Reply:
Nice. Thanks.